tensorflow/compiler/mlir/lite/ir/tfl_ops.cc

```diff
@@ -1444,12 +1444,50 @@ void FakeQuantOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
 // TODO(b/133486129): Implement shape inference for unpack
 
 static LogicalResult Verify(UnpackOp op) {
-  // TODO(antiagainst): Implement other checks as in
-  // tensorflow/lite/kernels/unpack.cc
+  if (op.getOperation()->getNumOperands() != 1)
+    return op.emitOpError("input count shoule be equal to 1");
 
   if (op.getOperation()->getNumResults() != op.num())
     return op.emitOpError("output count should match 'num' attribute");
 
+  auto input_type = op.input().getType().dyn_cast<ShapedType>();
+  if (!input_type.hasRank()) {
+    // If input has unknown rank, skip the checks.
+    return success();
+  }
+
+  if (input_type.getNumElements() <= 0)
+    return op.emitOpError("number of elements in input shoule be larger than 0");
+
+  const int64_t rank = input_type.getRank();
+  if (rank <= 0)
+    return op.emitOpError("input should be of rank larger than 0");
+
+  int64_t axis_value = op.axis().getSExtValue();
+  if (axis_value < 0)
+    axis_value += rank;
+  if (axis_value < 0 || axis_value >= rank)
+    return op.emitOpError()
+            << "op attribute 'axis' should be in range [-rank, rank), "
+            << "got rank = " << rank
+            << ", and axis = " << op.axis().getSExtValue();
+
+  llvm::SmallVector<int64_t, 4> output_shape;
```

Comment:
Could you extract these into a type op interface's build method? (https://github.com/tensorflow/tensorflow/blob/001ec7efbed18e9581e859513c5acc76e5aabbe9/tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc#L905) It results in generating more convenient build method and then also verifies the return type (https://github.com/tensorflow/tensorflow/commit/7779f4bbe0f50a77084561fe4308e27f45ec7a2b is an example but has other parts there).

If you can't get that to work, just add a TODO for me here.

Reply:
Thanks for pointing out this! It's very powerful. But I have a quick question here. If input is unranked, then how do we infer return type? For `UnpackOp`, the output is variadic, we cannot even know the number of results when the input is unranked. Is there any trick that we can achieve it? Thanks.

Reply:
Good question. So the number of outputs are always known (https://github.com/tensorflow/tensorflow/blob/acbfab2b0191ec7a845eb967b39f1cb08d3a3c3a/tensorflow/core/ops/array_ops.cc#L445) their shapes may not be (well, would not be in unranked case). I believe we keep that attribute still, else we would not be able to :-)

Reply:
Thank you for the clarification 😃 Updated.